lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -253,51 +435,360 @@ void FileEmitter::emitInstance(llvm::json::OStream &json,
+    structNameHint.resize(structNameHintLen);
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "- " << variable->name << ": " << emitted << "\n");
+  if (emitted) {
+    json.attributeBegin("value");
+    json.rawValue([&](auto &os) { os << emitted.expr; });
+    json.attributeEnd();
+    json.attribute("type_name", emitted.type.name);
+    if (auto dims = emitted.type.emitPackedDims(); !dims.empty())
+      json.attribute("packed_range", std::move(dims));
+    if (auto dims = emitted.type.emitUnpackedDims(); !dims.empty())
+      json.attribute("unpacked_range", std::move(dims));
+  }
+
+  json.objectEnd();
+}
+
+/// Emit the DI expression necessary to materialize a value.
+EmittedExpr FileEmitter::emitExpression(Value value) {
+  if (auto blockArg = dyn_cast<BlockArgument>(value)) {
+    auto module = dyn_cast<hw::HWModuleOp>(blockArg.getOwner()->getParentOp());
     if (!module)
-      module = defOp->getParentOfType<hw::HWModuleOp>();
-    if (module) {
-      if (auto arg = dyn_cast<BlockArgument>(value)) {
-        portName = dyn_cast_or_null<StringAttr>(
-            module.getInputNames()[arg.getArgNumber()]);
-      } else if (auto wireOp = value.getDefiningOp<hw::WireOp>()) {
-        portName = wireOp.getNameAttr();
-      } else {
-        for (auto &use : value.getUses()) {
-          if (auto outputOp = dyn_cast<hw::OutputOp>(use.getOwner())) {
-            portName = dyn_cast_or_null<StringAttr>(
-                module.getOutputNames()[use.getOperandNumber()]);
-            break;
-          }
-        }
+      return {};
+    auto name = module.getInputNameAttr(blockArg.getArgNumber());
+    if (!name)
+      return {};
+    return {JObject({{"sig_name", name.getValue()}}), value.getType()};
+  }
+
+  auto result = cast<OpResult>(value);
+  auto *op = result.getOwner();
+
+  // If the operation has only this one result and is named in some form, use
+  // that name.
+  if (op->getNumResults() == 1) {
+    // If a `hw.verilogName` is available, emit the value as just a reference to
+    // that name.
+    if (auto name = op->getAttrOfType<StringAttr>("hw.verilogName");
+        name && !name.empty())
+      return {JObject({{"sig_name", name.getValue()}}), result.getType()};
+
+    // Use the "name" attribute of certain Verilog-visible ops directly.
+    if (auto name = op->getAttrOfType<StringAttr>("name");
+        name && !name.empty() &&
+        isa<hw::WireOp, sv::WireOp, sv::RegOp, sv::LogicOp>(op))
+      return {JObject({{"sig_name", name.getValue()}}), result.getType()};
+  }
+
+  // Emit references to instance ports as `<instName>.<portName>`.
+  if (auto instOp = dyn_cast<hw::InstanceOp>(op)) {
+    auto instName = instOp->getAttrOfType<StringAttr>("hw.verilogName");
+    if (!instName)
+      instName = instOp.getInstanceNameAttr();
+    if (!instName)
+      return {};
+    auto portName = instOp
+                        .getPortList()[instOp.getPortIdForOutputId(
+                            result.getResultNumber())]
```

Comment:
Do we need correct verilog name for ports here? If so we need to look up a correct name using getPortVerilogName (c.f. https://github.com/llvm/circt/blob/dfc63c1bcedec0ef1ab5d45073e20ffd96025520/lib/Conversion/ExportVerilog/ExportVerilog.cpp#L4942-L4943)

Reply:
Excellent point, thanks! :+1: